### PR TITLE
Detect reasons of workflow job failures

### LINKED ISSUE #22: Gather data about the reason of workflow run failure
When a workflow fails, detect the type of failure:

* Network problem (hub.docker.com or apt/yum repository unavailable)
* Git problem (submodule can't be fetched)
* Dependency problem (these can be various)
* Test failure
* Test-run problem
* Build failure
* Cancelled by subsequent workflow on the same PR
* Cancelled manually
---

Subtasks

* 

- [ ] Write output to CSV, `--write csv`
- [ ] Filter workflows by date, `--since 2022-07-01`

---

Things to consider:

* This data will go to the test results database in #23 
* When we learn to detect the new type of failure, can we re-parse logs?
* If type of failure can't be detected, we should know about it somehow. Send a chat notification?
* It can be useful to split job steps https://github.com/tarantool/tarantool-qa/issues/253

### PR DESCRIPTION
**Pull request by artembo**
Detect reasons of workflow job failures
    
* Add regular expressions for about 30 most popular reasons.
* Calculate and print overall stats with `--failure-stats`.
* Log all found failures of a certain type with `--watch-failure`.
* Add documentation on new commmands and detectable failure reasons. Part of documentation is generated from code with `./multivac/docs.py`.
    
Resolves #22

### CODE REVIEW

File: multivac/gather_job_data.py [hunk 1]
```diff
@@ -3,12 +3,264 @@
...
+import sys
+from pprint import pprint
 
 PROJECT_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
 sys.path.append(PROJECT_DIR)
 
+version_matcher = re.compile(r"Current runner version: "
+                             r"'(\d*.\d*.\d*)'")
+
+job_was_newer_picked_up = re.compile(r'.*Job is about to start running on the runner.*')
```
Comment (Totktonada) [inline multivac/gather_job_data.py:16]: How about defining possible situations in a declarative way?

```python
situations = [
    {
        "line_re": <...>,
        "counter_name": <...>,
    },
    # <...>
]
```

It'll allow to get rid of the large if-elseif-<...> chain and will ease future extending of the list of situations.

Comment (NickVolynkin) [reply multivac/gather_job_data.py]: @Totktonada I've rebuilt this part, but in fact it's not enough. Some types of failures cannot be detected just from logs alone. Maybe I'll rewrite it with some kind of a sensor class.

Comment (Totktonada) [reply multivac/gather_job_data.py]: Thanks! I agree, it is not a full solution, but it structurizes a part of the code better. Most or at least many of failure types are detecteable by a regex on logs.

An idea with a sensor (or sensors) sounds good: it feels close to how I imagine the project architecture.

File: docs/gather_job_data.rst [hunk 2]
```diff
@@ -39,4 +44,53 @@ To gather data from a number of most recent workflows, use `--latest`:
...
+    dependency_autoreconf 1
+    tap_test_failed 1
+
+
+Command `--watch-failure name` will return a list of jobs where the named failure has been detected,
+along with the links to workflow runs on GitHub and matching log lines:
+
+..  code-block:: console
+
+    $ ./multivac/gather_job_data.py --latest 1000 --watch-failure testrun_test_failed                                                                          artembo/detect-errors ● ?  21:26:21
```
Comment (lastoCHka42) [inline docs/gather_job_data.rst:79]: Some trash from clipboard?

Comment (NickVolynkin) [reply docs/gather_job_data.rst:79]: exactly!

Comment (NickVolynkin) [reply docs/gather_job_data.rst]: fixed :)

File: multivac/gather_job_data.py [hunk 3]
```diff
@@ -7,19 +7,70 @@
...
 
 version_matcher = re.compile(r"Current runner version: "
                              r"'(\d*.\d*.\d*)'")
 
 
+# As far as the failure occurs at the end of the log, let's
+# start to parse the file from the end to speed up the process
+def reverse_readline(filename, buf_size=8192):
+    """An iterator that returns the lines of a file in reverse order"""
+    with open(filename, encoding='utf8') as fh:
```
Comment (lastoCHka42) [inline multivac/gather_job_data.py:23]: It is hard to understand what the `fh` means

Comment (NickVolynkin) [reply multivac/gather_job_data.py:23]: will fix in a further PR

**Pull request merged**